### DISCUSSION TITLE
Protected environments don't work when using tags for deployment

### DISCUSSION THREAD

**Discussion by WolfspiritM** [community/community#62991] [Actions] [answered]
### Select Topic Area

Bug

### Body

Hi,

We currently use github actions by creating a tag and a release from that and that starts a build and deploy based on the published event. However we are unable to protect our environment secrets to only be available for protected tags. I can only find protected branches.

Is that not possible or am I missing something?

When set to "protected branches" and creating a "v1.0.2" release for example we get:
Branch "v1.0.2" is not allowed to deploy to production due to environment protection rules.

EDIT: According to comments this was working before but broke with a latest "security" update by github and github is thinking about possible solutions.
So as far as I can see at the moment:
- If you use the Github Releases feature to create tags and build on that tags you can't use environments (at least not in a protected/secure way).
- If you want to actually use environments and "release" to specific environments you need to use branches (which in return kills the whole "Github Releases" feature as it doesn't know about branches).

More informations can be found [HERE](https://github.com/orgs/community/discussions/62991#discussioncomment-6690490)

**Comment by mehulrajdev** [comment]
https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment

this might help.
 mark as an answer

**Reply by WolfspiritM** [reply]
This doesn't help for tags.
There is no mention of tags or protected tags in there at all.

**Comment by jan-xyz** [comment]
We have the same problem. It worked until some days ago but suddenly it's failing. It seems that tags are no longer part of branch protection rules?

EDIT: Pretty sure that [this change](https://github.blog/changelog/2023-08-10-actions-runs-triggered-from-tags-or-forks-with-the-same-name-as-a-protected-branch-will-now-be-blocked/) broke it. It's nice that security is hardened around it, but is there also a way to still allow certain tags to access the credentials? Adding tag protection for example, to the list of possible targets.

EDIT 2: In [this thread](https://github.com/orgs/community/discussions/62991#discussioncomment-6738482) you can find more information and replies from the customer support.

EDIT 3: [this](https://github.com/community/community/discussions/62991#discussioncomment-7059887) thread has the official reply and announcement of protected tags.

**Reply by milotoor** [reply]
Interesting, thanks for digging that up. I'm still unsure though... we use a branch protection rule of `v*.*.*`, which doesn't match the name of any of our existing branches. Maybe I'm reading this wrong. Let's say someone's `prod` env had a branch protection rule so you could only deploy from the `main` branch. I can understand preventing deployments from a tag called `main`, that seems logical and enhances safety. But the implementation provided seems to just prohibit tag-based deployments to protected environments in all cases. Is that seriously the intent?

**Reply by WolfspiritM** [reply]
I think that the intent never was to let environments run on tags but it somehow was working with the way it was implemented. They never mentioned tags anywhere in the docs but seems like no one cared as it was just working when configuring branch protection rules. But with the rework done for the things mentioned in the blog post it just stops working for tags.

For me it feels like the github team doing github actions + environments didn't ever talk to the team doing github releases to make sure both work together...

**Reply by cortexcompiler** [reply]
We had been happily deploying using tags and the same `v.*.*.*` protection rule.  Now we are unable to restrict our deployment to those tags.  Feels like a step backwards.  This was the best way we could figure out to be able to deploy specific versioned artifacts corresponding to the tagged version.  Ideally we wanted a GitHub Actions input type that would allow us to choose a tag, similar to when dispatching a workflow
![image](https://github.com/community/community/assets/12128840/1fd0da6f-3d4c-4142-a4b8-0b29300c272a)

**Reply by Elliot-Alexander** [reply]
Also have been using this happily for 8 months. Now not working. Very frustrating and have no way of protecting production deployments while allowing for rollbacks to older tags.

**Reply by jan-xyz** [reply]
@WolfspiritM does it maybe make sense to update the label to Product Feedback to get GitHub's attention?

**Reply by george-zubrienko** [reply]
> @WolfspiritM does it maybe make sense to update the label to Product Feedback to get GitHub's attention?

I think they have it. I've been referred to this thread by support :) they also mentioned this change affects them internally and they are thinking about possible solutions

**Reply by jan-xyz** [reply]
Good to hear! I also approached our internal platform team to reach out to the enterprise support but haven't heard back yet, so this is the best news I got on this topic, yet! :D

**Reply by WolfspiritM** [reply]
@jan-xyz Sadly can't change the tags as far as I can see. Otherwise I'd change it at least to "bug" instead of question.

EDIT: I changed the main post a bit to give a bit more informations and also tried to change Question to Bug there but not sure if that will be picked up somehow after initial creation (I guess it won't)

**Reply by esilvia** [reply]
I wanted to refresh this convo. We also had a build-and-deploy workflow broken by this change that was previously working a month ago. Very similar scenario to the OP. 

1. Code is in branch `main`. 
2. Workflow triggers from a tag matching pattern `v1.*.*`.
3. Repo has protected tags rule for `v1.*.*`
4. dev Environment with "Deployment branches" set to "Protected Branches" (which includes `main`) works fine. This environment's builds are triggered by a push event, not by a tag event.
5. test Environment with "Deployment branches" set to "Selected Branches" with rule for `v1.*.*` (which "currently applies to 0 branches") no longer works with the aforementioned error about environment protection rules.

**Comment by comp615** [comment]
We also are experiencing this issue. All our production deploys started failing unexpectedly with branch protection issues. Previously, we were able to limit these by using a tag pattern, now we can't even use protection on "main" and deploy from a tag on main.

**Comment by fishpen0** [comment]
~~We noticed the regression started around the time rulesets came out of beta:
https://github.blog/changelog/2023-08-02-github-actions-required-workflows-will-move-to-repository-rules/
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/about-rulesets~~

This: https://github.com/orgs/community/discussions/62991#discussioncomment-6690490

**Comment by milotoor** [comment]
Same problem in multiple repositories. Our usual flow is to run a workflow with the ref set to a tag. The tag's name must match a pattern which we'd defined using branch protection rules. This worked as recently as two days ago, but began failing for us yesterday. This is a major disruption to our workflow and without having received any sort of notification of this change I have to assume it's a regression. To continue with deployments for now we've simply removed our branch protection rules... not great!

**Reply by mrjomp** [reply]
Same over here, no warnings were displayed to help us prevent this and instead from one day to the other deployments started crashing

**Comment by Paul-Gallagher** [comment]
👍 causing us grief too

**Comment by WolfspiritM** [comment]
As @jan-xyz pointed out this seems to be on purpose according to their blog post as obviously tags are no branches. However I have the feeling they're missing the fact that Releases are based on Tags and not on Branches. The missing of "protected tags" when it comes to environment protection seems to be completly ignoring the fact that a Release in Github doesn't actually have a branch.

So as far as I can see right now:
- If you use the Github Releases feature to create tags and build on that tags you can't use environments (at least not in a protected/secure way).
- If you want to actually use environments and "release" to specific environments you need to use branches (which in return kills the whole "Github Releases" feature as it doesn't know about branches).

I think there are two more options required:
- Allow to run environment deployments only on "release" action runs (not even on tags)
- Add a way to specify protected tags instead of protected branches for environments

**Reply by hanseartic** [reply]
There's a third option: installing a GH app, that would check if the tag is matching the specified filter and have that app approve/greenlight the deployment.

Right now there is no such app in the marketplace, though.

**Comment by nilsguillermin** [comment]
I don't understand, what is the workflow Github expects us to use with this change? Are we supposed to not be able to use protected environments with Actions at all?

**Reply by milotoor** [reply]
It seems like you can use protected environments with Actions, just not with tags (and therefore not with Releases). You may have to use a branch-based deployment strategy instead. It seems like a boneheaded decision to me.

**Comment by wokkaflokka** [comment]
Publicly commenting to agree with the others here. We use protected environments and automate releases using tags.

This change has led to a CI/CD outage at our company, and we currently are unsure how to maintain our existing CI/CD flow with this new behavior.

**Reply by tschm** [reply]
Same problem here.

**Comment by cortexcompiler** [comment]
We are also broken due to this.  We will now need to remove standard protection rules from environments in all of our repositories.  We were already working around a limitation in GitHub Actions where there is no option to specify an input parameter that can be a "tag" type and allow selecting a tag version.  So the `workflow_dispatch` trigger where you can select a specific tag was the best workaround we had come up with.  Feels like a warning with some notice or even monitoring how many folks this would break before introducing a breaking change would have been prudent.

How can we enforce that we are deploying from specific tag patterns?

**Comment by shkarface** [comment]
Same problem here, all our production deployments are at halt since there is no other alternative to this.

**Reply by hanseartic** [reply]
We are thinking now about introducing another GH workflow that would create a relaese-branch from a tag (that was created from a release) and than trigger the relase-workflows in those branches.

**Reply by shkarface** [reply]
This is a good workaround, but it's not easy to apply it to 250+ repos

**Reply by cortexcompiler** [reply]
Having to create branches to deploy from seems a step backwards in maturity, and is essentially jumping through extra hoops to work around a bad design from GitHub.  GitHub should be looking at ways to support deploying versioned artifacts in order to encourage repeatable deploys.

**Reply by hanseartic** [reply]
You're definitely right. Nevertheless you can have a look at the workflows I published here, to achieve the hoop-hopping: 

https://github.com/hanseartic/temporary_release_branches

**Comment by george-zubrienko** [comment]
Adding `Selected branches` with `v*` filter worked since we have started using github, now one of the repos has broken and I got a reply from support this is going to break further due to https://github.blog/changelog/2023-08-10-actions-runs-triggered-from-tags-or-forks-with-the-same-name-as-a-protected-branch-will-now-be-blocked/

I'm all in for security, but the `v*` approach has been working fine and I do no understand why a security change should nuke the deployment workflow protection in a such a sneaky manner...

**Comment by frugoman** [comment]
If tagging is not/(was never?) the expected way to handle release workflows, then why is the recommendation here like this? https://docs.github.com/en/actions/publishing-packages/publishing-nodejs-packages#publishing-packages-to-the-npm-registry

Maybe I'm reading it wrong?

**Reply by nathanchapman** [reply]
In that example it's checking out the default branch e.g. `main`
Which is a lacking approach because additional commits could have gone to `main` after creating the release tag.

GitHub basically broke the "GitHub Flow" pattern with this change for anyone using Environments

**Comment by elebiodaslingshot** [comment]
How does github continue to to screw over everyone with these decisions.

**Comment by elebiodaslingshot** [comment]
What is even the point of releases anymore

**Comment by cateprNIH** [comment]
Started running this issue as well.

**Comment by psafont** [comment]
This breaks publishing tagged releases to https://pypi.org/project/XenAPI/

**Comment by JasonPubC** [comment]
We had issues deploying our "tags" because we had an environment rule setup and it started enforcing it because it was not a "branch".

We had to remove the rule to enable us to continue deploying but not ideal as we only want to deploy tags in production - it would be best if we could set pattern rules for "branches" OR "tags".

Appears that a lot of people are waking up to this issue in their setups, hope a fix gets applied soon.

**Comment by george-zubrienko** [comment]
So just you folks don't burn out, here is some positive info from our support conversation after we've done some log exchange

```
I have found some clues on the logs that could point to a different issue. I suspect that there might have been an issue that interfered with the protection rules check and that this got stuck.
I have asked engineering to investigate in that direction and let us know.
If that is the case, I believe you should be able to use any other rule besides v*. For example v2*.

I will let you know as soon as engineering has investigated that possibility and let you know of our next steps.
```

**Reply by psafont** [reply]
> If that is the case, I believe you should be able to use any other rule besides v*. For example v2*.

This is not the case for my project, it's using `v*.*.*`. The error would be much better if it listed the rules it didn't match and what do they mean, like "Doesn't match any protected branch ["v*.*.*", ...] because we're not in any branch (detached head mode)"

**Reply by george-zubrienko** [reply]
I agree with you, and I'll make sure to give them feedback once they fix what was working before, so next time we get feature improvement, not feature regression ;)

**Reply by cateprNIH** [reply]
> If that is the case, I believe you should be able to use any other rule besides v*. For example v2*.

I'm using the tag pattern `releases/release-*` with `*` being a timestamp it is quite different.

**Reply by george-zubrienko** [reply]
so I got a response which is quite sad... I'll just paste it here. They reverted the change for one of our affected repos, but well... useless now anyway.

```
On August 10, we released a change in GitHub Actions to block tags and forks from branches from matching deployment branch rules on an environment ([changelog](https://github.blog/changelog/2023-08-10-actions-runs-triggered-from-tags-or-forks-with-the-same-name-as-a-protected-branch-will-now-be-blocked/)). Since this change addressed an active security vulnerability, we did not announce the change ahead of time. Since releasing that change, we received reports from several customers that their deployment workflows were broken. In order to unblock customers, we are temporarily providing the option to revert to the previous behavior on a case-by-case basis. This can only be performed by the GitHub Support team; there is no self-serve option. Moreover, we do not recommend enabling the old behavior on public repositories since they will be vulnerable to a publicly known security issue. The security risk of re-enabling the old behavior is that if a workflow deploys to an environment with a deployment branch rule:

Any user with Write access to the repository can deploy to that environment (by pushing a tag matching the protection rule)
If the workflow is triggered on pull_request_target, any user who can fork the repository can deploy to that environment (by using a fork with a branch matching the protection rule)
The user can bypass any [branch protection rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches) using this vulnerability. Therefore, we will support this workaround until September 14. During that time, we ask that any customers using tags or forks to match deployment branch rules to migrate to a more secure configuration. If you are using tags to match a deployment branch rule, you can create an equivalent branch name and use the branch name instead of the tag name. For example, for release v1.2.3, create a new branch named releases/v1.2.3 for the deployment.
```

**Reply by Nosfistis** [reply]
What is even the point of protected tags if we cannot use them in deployment?

**Reply by milotoor** [reply]
That's very sad. I wish the Releases team and the Actions team could get in sync about this, because the features are more or less mutually incompatible now.

**Reply by george-zubrienko** [reply]
An update: GH support told us Engineering is working on a solution that enables release deploy from protected tags. This might come around mid-September, until then they are offering rollback of this security fix (which we do not care since we do not allow forks). But this is for private repos only I think

**Comment by nilsguillermin** [comment]
Github support has offered the workaround of deploying from a branch instead of a tag. This doesn't seem to work however, if I branch `release/v*` off of `main` and then try to generate a release by tagging `v*` (This is how the Releases page works), I still get the same error. Unless I'm not understanding how to actually perform the workaround?

**Reply by WolfspiritM** [reply]
The "workaround" they probably mean is creating a branch from main called release/v* and not using tags at all. That also means listening to commit/push events instead of release events in the workflow. Which also means releases are completly pointless at the moment.

**Reply by hanseartic** [reply]
You could do it like this: https://github.com/hanseartic/temporary_release_branches.

I chose to create temporary branches with the same - for the time of their existence ambigous - name as the release tag, so that links in the Actions UI will point to the tag after the branch has been deleted (e.g. https://github.com/hanseartic/temporary_release_branches/tree/v0.0.1).

**Reply by WolfspiritM** [reply]
@hanseartic
> You could do it like this: https://github.com/hanseartic/temporary_release_branches.

Interesting idea, but how do you secure the PAT token in this case? Maybe I'm missing something but couldn't just everyone (who has push rights of course) create a branch which is not protected, extract the PAT by running a pipeline that exports the PAT in a way and then use the PAT to create release branches and circumventing any environment protection rules?

**Reply by hanseartic** [reply]
Correct, the `PAT` could be extracted from the workflow - as in any other malicious workflow.  So I think you're right and as far as secret-protection is concerned this opens a other loophole.

**Reply by cortexcompiler** [reply]
Could the workflow just use the GITHUB_TOKEN if given the appropriate permissions?  
However, I still think this is a lot of work to workaround what is essentially the removal of a feature from GitHub.  I think deploying from version tags is a useful use case and creating a branch just to deploy seems counterproductive.

**Reply by hanseartic** [reply]
Pushing a new branch is possible when `${{ github.token }}` has `content: write` permissions. However that does not trigger other workflow runs (i.e. the branch-deployment would not fire off).

It's definitely an awkward approach. Especially the cleanup-job runs on every update of `deployment_status` as the actual status can inly be used inside the run but not as a trigger. This will produce a bunch of skipped runs for each deployment.

A slightly cleaner approach would be a full-fledged GH App, that is installed in the repo – or just a rollback of the change.

Cleanest would be to have protected deployment tags as well

**Comment by rokpu** [comment]
Experiencing the same issue. Tags only work when the Environments Deployment branches are set to `All branches`, which is not secure.
Would really love a fix for this

**Comment by tomiGoljevscek** [comment]
Same here, we had staging and production environments protected with tags, created on master branch. Nothing is working anymore with our streamlined flow.

Let's hope GH will revert this absurd change...

**Comment by piotrowski** [comment]
Same here, we keep secrets in Environment that should be access only from release branches, we do not want to have environment without access protection. Hopefully you will roll-back this change.

**Comment by lyaunzbe** [comment]
Chiming in for the org @Artblocks: experiencing this issue as well, as the changes break the release flow in our repo that involves using tagged gh releases with protected environments

**Reply by jakerockland** [reply]
+++

**Comment by theothermattm** [comment]
Also causing issues, wasted an hour of my day during an important deployment only to find out that this was changed under the covers while it was working happily for us for months. As a paying organization, we are not happy!

**Comment by emil-lindq** [comment]
Switched from Gitlab last week and just ran into this issue while setting up the pipeline, not too good...

**Comment by BPreisner** [comment]
I am considering this as a braking change without notice since our prod deployment failed because of this :(

**Comment by elebiodaslingshot** [comment]
Top post for 3 weeks and no feedback from github... 👍

**Reply by nilsguillermin** [reply]
I got feedback in my support ticket. There is a feature flag they can turn on for certain repositories or for your entire organization that will switch it back to the old behavior. This feature flag will be permanently turned off mid-September however.

In the meantime, they're still looking for a way to support tag-based deployments while maintaining their desired security posture.

**Reply by elebiodaslingshot** [reply]
Appreciate that. Ill pass on their flag. Just kicks the can down the road. Im just absolutely lost as to how enforcing a mutable branch is more preferable to a tag, that is still mutable for security reasons.

**Reply by iqueryable** [reply]
Great info @nilsguillermin, but I too will pass until a more permanent solution is in place.

**Comment by levivannoort** [comment]
Experiencing the same issue, our deployment to production is also broken.

**Comment by nielsvanvelzen** [comment]
Just discovered this issue while publishing a new version of a library. We have the secrets for Maven Central in an environment scoped to the release branches and now I was forced to move those secrets to the repository scope and remove the environment. Not great.

**Reply by elebiodaslingshot** [reply]
#security

**Comment by nathanchapman** [comment]
Just received this response from support:

> After further discussion with our Engineering team, they are committing to having the new tag rule for deployments in place before we remove the workaround we have currently set on your organization. I do not have an exact ETA, but we are hoping to have these new tag rules in place in September.

**Comment by lelithium** [comment]
Also being affected by this. Looking for a workaround at the moment

**Reply by hanseartic** [reply]
you can find a workaround for the time being here: https://github.com/hanseartic/temporary_release_branches

**Reply by nilsguillermin** [reply]
You can also ask Github Support to switch it back to the old behavior via feature flag for the time being, though they will turn that off mid-September they say.

**Reply by VincentVerweij** [reply]
> You can also ask Github Support to switch it back to the old behavior via feature flag for the time being, though they will turn that off mid-September they say.

What are support saying about this? I've seen in a reply to this discussion that they are impacted as well internally. Do they acknowledge that this is a breaking change? Disabling something like that feature flag when many people are complaining here seems as if they do not acknowledge that this is a problem at all.

Edit: just read [this](https://github.com/community/community/discussions/62991#discussioncomment-6844269) reply. Let's hope they make something available this month...

**Comment by ilfa** [comment]
We use tags to [automate the deployment of the Dart library](https://dart.dev/tools/pub/automated-publishing#hardening-security-with-tag-protection-rules-on-github). And the only way to do it is to use tags. Hope GitHub team will figure out a solution that will allow to save protection for tags too.

**Comment by bdowne01** [comment]
Just chiming with a "me too".   Our Org was affected by this as well. 

We use environment branch protection rules to ensure only proper releases were deployed and not anything from an unexpected branch.   We had to remove those for now -- which is not great.

**Comment by b-urb** [comment]
This change and the lack of communication of such a breaking change makes no sense. The silence in this thread from anyone who is responsible is honestly unprofessional.

**Comment by george-zubrienko** [comment]
Hello Github,
While for our private org we were able to get an exception, we are also maintaining a public repo where we have no access to repo management. Now we cannot release a new version to PyPI due to this amazing security change, so I have to chase down the repo owner and leave people without bugfixes for a while. 

https://github.com/lidatong/dataclasses-json/issues/477

Maybe you conjure some public response finally, instead of behaving like classic MS-owned entity with no soul and no care about its users?

**Reply by rkustner** [reply]
@george-zubrienko did you get an exception from GitHub support or did your enterprise admin change a setting?

**Reply by george-zubrienko** [reply]
> @george-zubrienko did you get an exception from GitHub support or did your enterprise admin change a setting?

We got an exception through a ticket, I am not aware of such a setting. Support told its only possible via a ticket

**Comment by dubb-b** [comment]
Github, 

This is something you should address please.

**Comment by kaamoss** [comment]
Our production workflow is also affected, and by that I mean completely bricked unless we remove all protections, by this poorly communicated and implemented change.

**Comment by jeffrey-incentify** [comment]
Also having a problem with this. Can we get some help with this issue? Anyone get any sort of news / response from support?

**Comment by KateMateorova** [comment]
I agree, it's a big inconvenience!

**Comment by kevinyang77** [comment]
Also affected this. Please address this issue.

**Comment by EthanEFung** [comment]
_[removed by moderator]_ do either of you have connections to someone at Github who can respond?

**Comment by jappleba** [comment]
me as well

**Comment by cortexcompiler** [comment]
I want to add one more comment to this discussion.  As far as I can tell, the change to only support branch names in the environment protection rules forces teams to deploy to environments from branches.  We are using versioned deployable artifacts, which I see as a safer more mature way to manage deployments.  To me this change by GitHub seems to **force folks to use a less mature deployment pattern** (or forgo environment protection rules).  I hope the GitHub dev team is looking at ways to support this safer more mature approach.  

Keep in mind that it is **good practice to version your pipeline along with your code**, since structural changes in code may change how the pipeline works, this to me is a benefit of colocating your pipelines in your code (with GitHub Actions).  This change appears to force folks to disconnect their code versions from the pipeline that is deploying it, unless they use environment branches.

**Comment by george-zubrienko** [comment]
Alright a bit more clarify on when this is gonna resolve... I lack some emojis though to express my deep sadness.

> Just an update. The team is planning for beginning of October to release the new feature and will give two weeks grace period. At the same time more information will be provided on how to use and migrate to the new feature. If further extension is needed at that time we can discuss it again with the team.

New feature in this case is supporting deployment to specific environments from **protected tags**. Grace period refers to an exemption from this security fix you can get on the Org account if you get angry enough.

**Reply by jappleba** [reply]
can you link to that discussion please?

**Reply by milotoor** [reply]
I believe it's from a support ticket, so likely not linkable.

**Comment by milotoor** [comment]
Unclear why GitHub wouldn't link to this themselves, but they posted a public response yesterday including a roadmap for addressing the issue discussed here:

[Actions: More secure deployments to protected environments](https://github.com/github/roadmap/issues/825)

Quoting the conclusion:

> Admins who want to have more controlled deployments can **_NOW_** prevent self-reviews, configure Tag patterns to say, allow only Releases/* Tags to deploy to their Production Environment. And also configure that if a reviewer doesn't approve a deployment beyond 5 days, then the deployment would just fail.

Emphasis mine. Despite that, I'm still not seeing any ability to prevent self-reviews or configure tag patterns within my protected environment. Curious if anyone else is.

**Comment by N-Usha** [comment]
Thanks for all the patience on this folks. 🙏🏼 
Happy to share that GitHub team is now prioritising a few [enhancements around GitHub Actions - Protection rules](https://github.com/github/roadmap/issues/825) to enable admins to secure their deployment rollouts to their protected environments.

We are also expediting the support for protecting your deployment rollouts based on selected tag patterns and are tentatively planning to ship the feature in the next few weeks. We will update this discussion thread as the feature is rolled out along with a change blog.

> Allow deployments only for selected Tag patterns - Previously we supported [Deployment Branches](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#deployment-branches) with "Selected Branches" option. We are now enhancing this feature for controlling deployments to an environment based on “selected tag patterns”.

**Reply by jan-xyz** [reply]
Thanks for the update! I'm looking forward to it!

**Reply by theothermattm** [reply]
Thank you for acknowledging!!

**Reply by kyrofa** [reply]
Excellent, just got bit by this today.

**Comment by ajaykarthikr** [comment]
This issue has been resolved today. 

https://github.blog/changelog/2023-10-06-actions-secure-deployment-rollouts-to-protected-environments-based-on-select-tag-patterns/

**Reply by kyrofa** [reply]
Thank you, @ajaykarthikr! Can I ask for one small addition? A "protected branches and tags" option here so we don't need to duplicate all our rules?

![Screenshot from 2023-10-07 06-45-43](https://github.com/community/community/assets/946674/7e539d49-cc3c-4987-b42c-122dd83c9124)

**Reply by N-Usha** [reply]
Thanks for that request @kyrofa . We are actively evaluating an integration of  "Deployment branches and tags" feature with  [Repository Rulesets](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/about-rulesets#about-rulesets-protected-branches-and-protected-tags) feature to support both "Protected branches" and " Protected Tags" eventually.  And we believe that will help our users to centrally manage all their "rules" around branches & tags centrally at both repo and Organizational level, which would also be honoured by Environments for controlling Deployment rollouts.
Hope that helps.

We'll keep you posted when we have some timelines to share. Thanks!

**Comment by brcrista** [comment, marked as answer]
:wave: I'm here from the GitHub Engineering team.  We've been following this thread and we understand this change meant you couldn't ship your code to production.  That's something we take seriously.  I wanted to explain our reasoning with this change and share what we've done to try to make it better.

Tags worked with deployment branch patterns because of a bug.  Unfortunately, this wasn't secure. Even with protected tags, anyone who could write to or fork the repo could bypass the protection rule.  The breaking change, in our opinion, was the better of two bad choices.  At the same time, we started work on a new feature to provide a secure implementation of this functionality.

So today, we released [deployment tag protection rules](https://github.blog/changelog/2023-10-06-actions-secure-deployment-rollouts-to-protected-environments-based-on-select-tag-patterns/).  They let you specify changes that can run against environments through tag patterns.  Unlike the old behavior of the deployment branches rule, these patterns will only match tags in your repo.

You can use this script to help migrate from branch to tag patterns through the API:

```sh
REPO_ADMIN_PAT="<personal access token with the repo:admin scope>"
REPO_NWO="<name of the repository, like user/repo>"
ENVIRONMENT="<name of the environment to create the rule>"

# create tag policy
curl -L -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Bearer $REPO_ADMIN_PAT" https://api.github.com/repos/$REPO_NWO/environments/$ENVIRONMENT/deployment-branch-policies -d '{"name":"v**", "type":"tag"}' 

# list all
curl -L -H "Accept: application/vnd.github+json" -H "Authorization: Bearer $REPO_ADMIN_PAT" https://api.github.com/repos/$REPO_NWO/environments/$ENVIRONMENT/deployment-branch-policies 
```

We're continuing to work on more robust security policies for building and delivering code.  Please let us know your feedback!

**Reply by jan-xyz** [reply]
It worked out of the box and as expected! 

One small note: the drop-down box to pick between branch and tag pattern doesn't work in Safari. The drop-down just doesn't show anything but branch. Our workaround is to tell the teams to use Firefox for now to add tag patterns.

![image](https://github.com/community/community/assets/5249233/0220ccc8-b09d-43ec-8bba-f6df9be21f70)

**Reply by jmarshall** [reply]
Indeed re Safari, which I reported three days ago as #69439 and ticket 2364619.

Retesting it now, it appears to be working now in Safari.

**Reply by Aeolun** [reply]
Docs need some updates. It looks like the example responses for '[List deployment branch policies](https://docs.github.com/en/rest/deployments/branch-policies?apiVersion=2022-11-28#list-deployment-branch-policies) and [Get a deployment branch policy](https://docs.github.com/en/rest/deployments/branch-policies?apiVersion=2022-11-28#get-a-deployment-branch-policy) do not show the `type`, which led me to believe they didn't exist.

I think this also means the endpoints need to be renamed to `branch/tag` instead of calling it branch for both options.

**Reply by bayer-alvaro** [reply]
Thanks for your details , but having check validity of my regex, I still see error
`The deployment was rejected or didn't satisfy other protection rules.
`
And I've checked this pattern in https://regexr.com/ and https://rubular.com/, however despite of having several tags like 
<img width="229" alt="image" src="https://github.com/community/community/assets/119419511/f2aa9da7-22b0-4cba-b4f4-51a5af634fb4">

protection rules don't recognize them as valid ? :(

<img width="397" alt="image" src="https://github.com/community/community/assets/119419511/19561557-45db-4a20-b237-9421659be728">

**Reply by t3chguy** [reply]
@bayer-alvaro the format isn't regex, its https://ruby-doc.org/core-2.5.1/File.html#method-c-fnmatch as per https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-tag-protection-rules#adding-tag-protection-rules

**Reply by bayer-alvaro** [reply]
Yep. Thank you. I finally realized and reduced my matching pseudo-regex. It's solved now. Thx.

**Reply by axi92** [reply]
Can you share your pattern for semver?

**Comment by theothermattm** [comment]
Just want to point out that the environments feature doesn't work particularly well with the `schedule` function since it limits you to running an action only on the default branch.  

I have more details [in this comment](https://github.com/orgs/community/discussions/26374#discussioncomment-7242461), since it's more related to the schedule function than this feature (I think!).

**Comment by nicktalb** [comment]
This was working for a few weeks but now GitHub have broken it AGAIN. I deployed a GitHub release with tag protection rules set to `v*` fine yesterday afternoon. This morning another release is failing with:
> Branch "main" is not allowed to deploy to production due to environment protection rules.

Seems like GitHub is now passing the wrong ref to the pipeline. How hard is it to get these basic things right?!

Update: I've just noticed the branch instead of the tag is now showing up in the Actions deployments page, there has been a change to the information passed to the pipeline between yesterday afternoon and this morning:
![image](https://github.com/community/community/assets/3307312/4d97aaab-7c29-46f5-a3c0-0e4060016d16)

**Reply by comp615** [reply]
Also seeing this behavior!

**Comment by nicktalb** [comment]
Broken again GitHub. Almost a year to the day... Please fix asap!
> Branch "main" is not allowed to deploy to production due to environment protection rules.

**Comment by samus-aran** [comment]
Hi y'all. Thank you all for your feedback and engagement on this topic! Since this discussion was opened, the feature [deployment tag protection rules](https://github.blog/changelog/2023-10-06-actions-secure-deployment-rollouts-to-protected-environments-based-on-select-tag-patterns/) has been shipped. As a result, the original issue discussed here has been addressed.

Due the low activity since this, we are closing this discussion. If you encounter any incidents or have further questions regarding protected tags or related features, we encourage you to create a new discussion. This ensures we can provide focused support and track any new issues effectively. 👍
